crates/flux-syntax/src/grammar.lalrpop

```diff
@@ -543,7 +543,35 @@ Level10: surface::Expr = {
             span: cx.map_span(lo, hi),
         }
     },
-    "(" <Level1> ")"
+    Constructor,
+}
+
+Constructor: surface::Expr = {
```

Comment:
@nilehmann , I'm not sure how to include this into `Expr` without blowing up the Parser. Seems like the issue is coming from the rule for refinements like `new_x { new_x == .... }`. Maybe I should do as you say and collapse refinements into exprs?

Reply:
oh, I was fearing this. Did some testing. The following works

```
Level10: surface::Expr = {
    "{" <f:Ident> ":" <arg:Level1> "}" => {
        todo!()
    },
    <lo:@L> "if" <p:Level1> "{" <e1:Level1> "}" <e2:ElseIf> <hi:@R> => {
        ...
    },
    ...
```    

but the following is ambiguous

```
Level10: surface::Expr = {
   <name:Ident>  "{" <f:Ident> ":" <arg:Level1> "}" => {
        todo!()
    },
    <lo:@L> "if" <p:Level1> "{" <e1:Level1> "}" <e2:ElseIf> <hi:@R> => {
       ...
    },
    ...
``` 

The issue is `if` expressions when a constructor is used in the condition, e.g.,

```rust
if S {f: 0} { ... } else { ... }
```    
When looking at the `{` after `S` it's not possible (without lookahead) to know whether `S` is a variable and `{` starts the `then` branch or whether `S ` is a constructor. Writing the same in rust would tell you to disambiguate with parentheses.

Reply:
The solution in `syn` is to disallow structs in some contexts. This is not very pretty, but I don't think there is a way around it. We could something similar by parameterizing the parser with a boolean. Something like:

```
pub Expr = Level1<"true">;

Level1<AllowStruct> = NonAssoc<BinOp1, Level2<AllowStruct>>;  // <=>
Level2<AllowStruct> = LeftAssoc<BinOp2, Level3<AllowStruct>>; // =>
Level3<AllowStruct> = LeftAssoc<BinOp3, Level4<AllowStruct>>; // ||
Level4<AllowStruct> = LeftAssoc<BinOp4, Level5<AllowStruct>>; // &&
Level5<AllowStruct> = NonAssoc<BinOp5, Level6<AllowStruct>>;  // ==, !=, >=, <=
Level6<AllowStruct> = LeftAssoc<BinOp6, Level7<AllowStruct>>; // >>, <<
Level7<AllowStruct> = LeftAssoc<BinOp7, Level8<AllowStruct>>; // +, -
Level8<AllowStruct> = LeftAssoc<BinOp8, Level9<AllowStruct>>; // *, %, /
Level9<AllowStruct> = {
    <lo:@L> <op:UnOp> <e:Level10<AllowStruct>> <hi:@R> => {
        surface::Expr {
            kind: surface::ExprKind::UnaryOp(op, Box::new(e)),
            node_id: cx.next_node_id(),
            span: cx.map_span(lo, hi),
        }
    },
    <Level10<AllowStruct>>
}

Level10<AllowStruct>: surface::Expr = {
    <name:Ident> "{" <f:Ident> ":" <arg:Expr> "}" if AllowStruct == "true" => {
        todo!()
    },
    <lo:@L> "if" <p:Level1<"false">> "{" <e1:Expr> "}" <e2:ElseIf> <hi:@R> => {
        surface::Expr {
            kind: surface::ExprKind::IfThenElse(Box::new([p, e1, e2])),
            node_id: cx.next_node_id(),
            span: cx.map_span(lo, hi),
        }
    },
    ...
```

Reply:
I'd still start by refactoring `fhir` to collapse `RefineArg` and `Expr` and submit that in a separate PR, otherwise, it'll get too big.

Reply:
Ok cool - good to know